File: windows/Helpers/TaskSorting.cs
using Hatch.Models;

namespace Hatch.Helpers;

public static class TaskSorting
{
    // CreatedAt keeps whatever kind its writer stamped — this app local, the companion
    // UTC — and DateTime compares raw ticks, so the two sources interleave hours apart.
    public static DateTime CreatedInstant(TodoItem task) => task.CreatedAt.ToUniversalTime();

    public static IEnumerable<TodoItem> NewestFirst(IEnumerable<TodoItem> tasks) =>
        tasks.OrderByDescending(CreatedInstant);

    // Important smart list: highest priority first, then newest-first within a tier.
    public static IEnumerable<TodoItem> ForImportant(IEnumerable<TodoItem> tasks) =>
        tasks.OrderByDescending(t => t.Priority).ThenByDescending(CreatedInstant);
}

File: mobile/shared/src/commonMain/kotlin/dev/hatch/sync/TaskSorting.kt
package dev.hatch.sync

import kotlin.time.Instant

// Transcription of windows/Helpers/TaskSorting.cs.
object TaskSorting {

    // Never compare createdAt as text: Windows stamps local time, this app stamps UTC, so
    // "18:00+07:00" and "11:00Z" are the same moment but sort seven hours apart.
    // Unparseable sorts oldest rather than jumping to the top.
    fun createdInstant(task: TodoItem): Instant =
        runCatching { Instant.parse(task.createdAt) }.getOrElse { Instant.DISTANT_PAST }

    fun newestFirst(tasks: List<TodoItem>): List<TodoItem> =
        tasks.sortedByDescending { createdInstant(it) }

    fun forImportant(tasks: List<TodoItem>): List<TodoItem> =
        tasks.sortedWith(
            compareByDescending<TodoItem> { it.priority }.thenByDescending { createdInstant(it) }
        )
}
